dbms/src/Flash/Statistics/ExecutorStatisticsCollector.cpp

```diff
@@ -65,22 +80,121 @@ void ExecutorStatisticsCollector::initialize(DAGContext * dag_context_)
                 TableScanStatistics,
                 TopNStatistics,
                 WindowStatistics,
-                ExpandStatistics>(executor_id, &executor))
+                ExpandStatistics>(&executor))
         {
             throw TiFlashException(
-                fmt::format("Unknown executor type, executor_id: {}", executor_id),
+                fmt::format("Unknown executor type, executor_id: {}", executor.executor_id()),
                 Errors::Coprocessor::Internal);
         }
         return true;
     });
+
+    fillListBasedExecutorsChild();
+}
+
+void ExecutorStatisticsCollector::fillListBasedExecutorsChild()
+{
+    if (!dag_context->return_executor_id)
+    {
+        // fill list-based executors child
+        auto size = dag_context->dag_request->executors_size();
+        RUNTIME_CHECK(size > 0);
+        const auto & executors = dag_context->dag_request->executors();
+        String child;
+        for (int i = 0; i < size; ++i)
+        {
+            const auto & executor_id = executors[i].executor_id();
+            if (i != 0)
+                profiles[executor_id]->setChild(child);
+            child = executor_id;
+        }
+    }
```

Comment:
I prefer to set child id in the constructor of `ExecutorStatistics`, like
```
if (dag_context->is_list_base_executor)
{
    std::optional<String> child_id;
	auto children = child_id ? {child_id.get()} : {};
    traverseExecutorListReverse(dag_context->dag_request, [&](const tipb::Executor & executor) {
        if (!append<
            AggStatistics,
            ExchangeReceiverStatistics,
            TableScanStatistics,
            TopNStatistics,
            WindowStatistics,
            ExpandStatistics>(executor_id, &executor, children))
        {
            throw TiFlashException(
                fmt::format("Unknown executor type, executor_id: {}", executor_id),
                Errors::Coprocessor::Internal);
        }
		child_id = executor_id;
        return true;
	};
}
else
{
    traverseExecutorTree(dag_context->dag_request, [&](const tipb::Executor & executor) {
        if (!append<
            AggStatistics,
            ExchangeReceiverStatistics,
            TableScanStatistics,
            TopNStatistics,
            WindowStatistics,
            ExpandStatistics>(executor_id, &executor, getChildren(&executor)))
        {
            throw TiFlashException(
                fmt::format("Unknown executor type, executor_id: {}", executor_id),
                Errors::Coprocessor::Internal);
        }
        return true;
    };
}
```

Reply:
IMO, the code for the same function should be in the same place.

Reply:
I prefer to move the add child process outside of the constructor to make it clear that we need to init every executors' children. And we don't need to add a 'if else' block here.

Reply:
If so, we should move https://github.com/pingcap/tiflash/pull/6995/files#diff-3aff7923729080b9f5821c18130031eca8fa0504b560e8a256893d05c778828fR48-R56 outside the constructor too?